databricks/koalas/frame.py

```diff
@@ -8331,6 +8331,82 @@ def quantile(self, q=0.5, axis=0, numeric_only=True, accuracy=10000):
 
         return DataFrame(internal) if not result_as_series else DataFrame(internal).T[key]
 
+    def query(self, expr, inplace=False):
+        """
+        Query the columns of a DataFrame with a boolean expression.
+
+        Parameters
+        ----------
+        expr : str
+            The query string to evaluate.
+
+            You can refer to column names that contain spaces by surrounding
+            them in backticks.
+
+            For example, if one of your columns is called ``a a`` and you want
+            to sum it with ``b``, your query should be ```a a` + b``.
+
+        inplace : bool
+            Whether the query should modify the data in place or return
+            a modified copy.
+
+        Returns
+        -------
+        DataFrame
+            DataFrame resulting from the provided query expression.
+
+        Examples
+        --------
+        >>> df = ks.DataFrame({'A': range(1, 6),
+        ...                    'B': range(10, 0, -2),
+        ...                    'C C': range(10, 5, -1)})
+        >>> df
+           A   B  C C
+        0  1  10   10
+        1  2   8    9
+        2  3   6    8
+        3  4   4    7
+        4  5   2    6
+
+        >>> df.query('A > B')
+           A  B  C C
+        4  5  2    6
+
+        The previous expression is equivalent to
+
+        >>> df[df.A > df.B]
+           A  B  C C
+        4  5  2    6
+
+        For columns with spaces in their name, you can use backtick quoting.
+
+        >>> df.query('B == `C C`')
+           A   B  C C
+        0  1  10   10
+
+        The previous expression is equivalent to
+
+        >>> df[df.B == df['C C']]
+           A   B  C C
+        0  1  10   10
+        """
+        if not isinstance(expr, str):
+            raise ValueError(
+                'expr must be a string to be evaluated, {} given'
+                .format(type(expr)))
+        if not isinstance(inplace, bool):
+            raise ValueError(
+                'For argument "inplace" expected type bool, received type {}.'
+                .format(type(inplace).__name__))
+
+        sdf = self._sdf.filter(expr)
```

Comment:
@itholic how does pandas handle multi-index columns?

Reply:
let me test!

Reply:
i just tested, they seems not support for multi-index columns.

i added the code for catching that exception.

Reply:
@itholic can you add the limitation of this API in the docstring? You could use `.. note::`.

- multiindex columns are not supported.
- internal columns that starting with `__`.. are able to access; however, they are not supposed to be accessed.
- this delegates to Spark SQL so the syntax follows Spark SQL

Reply:
Also, adds another note that, if you want to use multidex columns, people can easily work around by using `DataFrame.apply`.

Reply:
yes, i got it. thanks!! 👍

Reply:
> internal columns that starting with __.. are able to access; however, they are not supposed to be accessed.

I'm curious about this; if we add note about this, then should we add this to all APIs since all APIs are using internal columns??